src/NServiceBus.Core/Persistence/Learning/SagaPersister/SagaStorageFile.cs

```diff
@@ -40,40 +38,52 @@ public void Dispose()
             fileStream = null;
         }
 
-        public static bool TryOpen(Guid sagaId, SagaManifest manifest, out SagaStorageFile sagaStorageFile)
+        public static Task<SagaStorageFile> Open(Guid sagaId, SagaManifest manifest)
         {
             var filePath = manifest.GetFilePath(sagaId);
 
             if (!File.Exists(filePath))
             {
-                sagaStorageFile = null;
-                return false;
+                return Task.FromResult<SagaStorageFile>(null);
             }
 
-            sagaStorageFile = new SagaStorageFile(new FileStream(filePath, FileMode.Open, FileAccess.ReadWrite, FileShare.ReadWrite, DefaultBufferSize, FileOptions.Asynchronous), manifest);
-
-            return true;
+            return OpenWithDelayOnConcurrency(manifest, filePath, FileMode.Open);
         }
 
-        public static SagaStorageFile Create(Guid sagaId, SagaManifest manifest)
+        public static Task<SagaStorageFile> Create(Guid sagaId, SagaManifest manifest)
         {
             var filePath = manifest.GetFilePath(sagaId);
 
-            return new SagaStorageFile(new FileStream(filePath, FileMode.CreateNew, FileAccess.ReadWrite, FileShare.ReadWrite, DefaultBufferSize, FileOptions.Asynchronous), manifest);
+            return OpenWithDelayOnConcurrency(manifest, filePath, FileMode.CreateNew);
         }
 
+        static async Task<SagaStorageFile> OpenWithDelayOnConcurrency(SagaManifest manifest, string filePath, FileMode fileAccess)
+        {
+            try
+            {
+                return new SagaStorageFile(new FileStream(filePath, fileAccess, FileAccess.ReadWrite, FileShare.None, DefaultBufferSize, FileOptions.Asynchronous), manifest);
+            }
+            catch (IOException)
+            {
+                // give the other task some time to complete the saga to avoid retrying to much
+                await Task.Delay(100)
+                    .ConfigureAwait(false);
+
+                return new SagaStorageFile(new FileStream(filePath, fileAccess, FileAccess.ReadWrite, FileShare.None, DefaultBufferSize, FileOptions.Asynchronous), manifest);
```

Comment:
while I'm not against this approach, I'd still like to mention my counter-proposal to not do a single immediate retry but instead rethrow after the delay to force a regular immediate retry after the delay period. This approach would be cleaner when involving transports with lease/tx timeouts. The learning persistence is more vulnerable to concurrency issues because it now locks the file for quite a long time, increasing the chance of concurrent access, but at the same time, the 100ms seem to be an arbitrary number which easily won't hold up once you're going to do some real IO in your handler, therefore rendering this mechanism useless anyway?